Initializers allow initializing a param field with a non-param value

### Summary of Problem

```chapel
class C {
  param p: int;
  proc init() {
    var x = 7;
    p = x;
    super.init();
  }
}

var c = new C();
writeln(c);
delete c;
```

[edit 2/9 cassella -- simplify test case to no longer use a record. ]


```
fortytwo@magrathea:~/src/chpl/tfn$ chpl param-from-record.chpl 
In file included from /tmp/chpl-fortytwo-8306.deleteme/_main.c:28:0:
/tmp/chpl-fortytwo-8306.deleteme/ChapelIO.c: In function ‘writeThisFieldsDefaultImpl’:
/tmp/chpl-fortytwo-8306.deleteme/ChapelIO.c:237:13: error: ‘call_tmp_chpl’ undeclared (first use in this function)
 call_tmp7 = call_tmp_chpl;
             ^
/tmp/chpl-fortytwo-8306.deleteme/ChapelIO.c:237:13: note: each undeclared identifier is reported only once for each function it appears in
/home/fortytwo/src/chapel/runtime/etc/Makefile.exe:27: recipe for target '/tmp/chpl-fortytwo-8306.deleteme/param-from-record.tmp' failed
make: *** [/tmp/chpl-fortytwo-8306.deleteme/param-from-record.tmp] Error 1
error: compiling generated source
```

I'd have expected an error on the ``p = x`` line since ``x`` isn't a param.

[ edit again 2/9 cassella -- adjust the expected behavior too. ]

**Associated Future Test(s):**
None yet.
<!-- Are there any tests in Chapel's test system that demonstrate this issue?
     e.g. [`test/path/to/foo.chpl`](
           https://github.com/chapel-lang/chapel/blob/master/test/path/to/foo.chpl
          ) #1234 -->

### Configuration Information

```
fortytwo@magrathea:~/src/chpl/tfn$ chpl --version
chpl version 1.17.0 pre-release (ee5378d)
Copyright (c) 2004-2018, Cray Inc.  (See LICENSE file for more details)

fortytwo@magrathea:~/src/chpl/tfn$ printchplenv --anonymize
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: intrinsics
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_AUX_FILESYS: none

fortytwo@magrathea:~/src/chpl/tfn$ gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.6) 5.4.0 20160609
```

--- a/compiler/resolution/functionResolution.cpp
+++ b/compiler/resolution/functionResolution.cpp
@@ -4902,6 +4902,15 @@ static void resolveInitField(CallExpr* call) {
   if (t == dtUnknown)
     INT_FATAL(call, "Unable to resolve field type");
 
+  if (fs->hasFlag(FLAG_PARAM)) {
+    if (isLegalParamType(t) == false) {
+      USR_FATAL_CONT(fs, "'%s' is not of a supported param type", fs->name);
+    }
+    if (rhs->symbol()->isParameter() == false) {
+      USR_FATAL_CONT(call, "Initializing parameter '%s' to value not known at compile time", fs->name);
+    }
+  }
+
   if (t == dtNil && fs->type == dtUnknown)
     USR_FATAL(call->parentSymbol, "unable to determine type of field from nil");
   if (fs->type == dtUnknown) {

--- a/test/classes/initializers/generics/init-param-from-non-param.future
+++ b/test/classes/initializers/generics/init-param-from-non-param.future
@@ -1,3 +1,0 @@
-bug: initializers allow assignment of non-params to param fields
-#8406
-

--- a/test/classes/initializers/generics/init-param-from-non-param.good
+++ b/test/classes/initializers/generics/init-param-from-non-param.good
@@ -1,1 +1,2 @@
+init-param-from-non-param.chpl:3: In initializer:
 init-param-from-non-param.chpl:5: error: Initializing parameter 'p' to value not known at compile time

--- a/test/classes/initializers/generics/param-record-field.future
+++ b/test/classes/initializers/generics/param-record-field.future
@@ -1,6 +1,0 @@
-bug: classes allow param record fields
-#8422
-
-Without the writeln(), this compiles.
-With the writeln(), there's a failure to build the generated source in
-ChapelIO.c

--- a/test/types/records/init/record-with-record-param.future
+++ b/test/types/records/init/record-with-record-param.future
@@ -1,2 +1,0 @@
-error message: Confusing error message
-#8438

